src/intrusive.rs

```diff
@@ -15,21 +16,303 @@ use core::{
+        self.next_seq = self.last_three[2] + 1;
+    }
+}
+
+struct StorageListInner {
+    list: List<NodeHeader>,
+    seq_state: SeqState,
+}
+
+impl StorageListInner {
+    async fn get_or_populate_latest<S: NdlDataStorage>(
+        &mut self,
+        s: &mut S,
+        buf: &mut [u8],
+    ) -> Result<Option<u32>, LoadStoreError> {
+        if self.seq_state.has_hydrated {
+            let max = self.seq_state.last_three.iter().copied().max().unwrap_or(0);
+            if max == 0 {
+                return Ok(None);
+            } else {
+                return Ok(Some(max));
+            }
+        }
+
+        // We have NOT hydrated, do so now
+        let Ok(mut iter) = s.iter_elems().await else {
+            todo!()
+        };
+
+        let mut current: Option<u32> = None;
+        // todo: real crc
+        let mut crc = FakeCrc32::new();
+
+        'outer: loop {
+            let item = loop {
+                let res = iter.next(buf).await;
+                match res {
+                    // Got at item
+                    Ok(Some(Some(item))) => break item,
+                    // Got a bad item, continue
+                    Ok(Some(None)) => {}
+                    // end of list
+                    Ok(None) => break 'outer,
+                    // flash error
+                    Err(_e) => todo!(),
+                };
+            };
+
+            match item.data() {
+                Elem::Start { seq_no } => {
+                    current = Some(seq_no);
+                    // todo: real crc
+                    // todo: We should make sure next_seq is ALWAYS larger than any seen start,
+                    // to ensure we don't accidentally re-use a bad one.
+                    crc = FakeCrc32::new();
+                }
+                Elem::Data { data } => {
+                    if current.is_none() {
+                        continue;
+                    } else {
+                        // todo: real crc
+                        crc.update(data.key_val());
+                    }
+                }
+                Elem::End { seq_no, calc_crc } => {
+                    let check_crc = crc.finalize();
+                    crc = FakeCrc32::new();
+                    let Some(seq) = current.take() else {
+                        continue;
+                    };
+                    if seq != seq_no {
+                        continue;
+                    }
+                    // todo: real crc
+                    if calc_crc != check_crc {
+                        continue;
+                    }
+                    self.seq_state.insert_good(seq_no);
+                }
+            }
+        }
+
+        self.seq_state.has_hydrated = true;
+        let max = self.seq_state.last_three.iter().copied().max().unwrap_or(0);
+        if max == 0 {
+            self.seq_state.next_seq = 1;
+            Ok(None)
+        } else {
+            Ok(Some(max))
+        }
```

Comment:
This looks a bit weird to me...
Why do we set next_seq to 1?
Feels a bit magic this. If that's how an empty seq_state should be initialized, maybe that should be done at the actual point of initialization?

Reply:
Yeah, there's magic here. I'm using 0 to mean "None", really I should make it `Option<NonZeroU32>` or so